### PR TITLE
Implemented dispatcher that works like WPF one

### PR DESCRIPTION
**Pull request by kekekeks**
Ported some parts of WPF's Dispatcher and made the rest to be compatible with non-Windows platforms.

See #10520 for more details.

It should work without limitations with Windows/X11 and macOS. For other platforms it still uses IPlatformThreadingInterface.

Linux framebuffer and headless platform shouldn't have any issues either but are not implemented yet.

### Current limitations on other platforms:

**Foreground processing should come before input**

For WASM we'll have to check if dispatcher has asked for foreground processing _in every single JS->managed callback_, since it's not possible to request a callback to be inserted at the start of the event queue. It's likely the case for Android too, but we need to investigate further.

For iOS we can use the same CFRunLoopObserver approach as with macOS

**Background processing should come after input**

For Android it's possible to query if there are pending input events. There is a preview API for web as well. For iOS and legacy web browsers we can use deadline scheduling (already implemented) when we give up control if background jobs were running more than a small amount of time and reschedule processing via a timer.

### CODE REVIEW

File: src/Avalonia.Base/Threading/IDispatcher.cs [hunk 1]
```diff
@@ -26,30 +26,5 @@ public interface IDispatcher
...
         /// <param name="priority">The priority with which to invoke the method.</param>
         void Post(Action action, DispatcherPriority priority = default);
-
-        /// <summary>
-        /// Posts an action that will be invoked on the dispatcher thread.
-        /// </summary>
-        /// <param name="action">The method.</param>
-        /// <param name="arg">The argument of method to call.</param>
-        /// <param name="priority">The priority with which to invoke the method.</param>
-        void Post(SendOrPostCallback action, object? arg, DispatcherPriority priority = default);
```
Comment (workgroupengineering) [inline src/Avalonia.Base/Threading/IDispatcher.cs:36]: a curiosity, why did you remove this method? It was added to avoid allocating clousure.

Comment (kekekeks) [reply src/Avalonia.Base/Threading/IDispatcher.cs:36]: We could re-add it by introducing another specialized DispatcherOperation class, I guess.

File: src/Avalonia.Base/Threading/JobRunner.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ public JobRunner(IPlatformThreadingInterface? platform)
         /// <param name="priority">Priority to execute jobs for. Pass null if platform doesn't have internal priority system</param>
         public void RunJobs(DispatcherPriority? priority)
         {
-            var minimumPriority = priority ?? DispatcherPriority.MinValue;
+            var minimumPriority = priority ?? DispatcherPriority.MinimumActiveValue;
```
Comment (maxkatz6) [inline src/Avalonia.Base/Threading/JobRunner.cs:31]: Do we even need JobRunner/IJob after this PR?

File: src/Avalonia.Base/Threading/AvaloniaSynchronizationContext.cs [hunk 3]
```diff
@@ -30,7 +30,7 @@ public static void InstallIfNeeded()
         /// <inheritdoc/>
         public override void Post(SendOrPostCallback d, object? state)
         {
-           Dispatcher.UIThread.Post(d, state, DispatcherPriority.Background);
+            Dispatcher.UIThread.Post(() => d(state), DispatcherPriority.Background);
```
Comment (maxkatz6) [inline src/Avalonia.Base/Threading/AvaloniaSynchronizationContext.cs:33]: to not forget, this method should use overload that accepts state as a parameter instead of creating a closure

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032008-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032016-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032021-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032035-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032039-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032043-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032064-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: samples/ControlCatalog/App.xaml [hunk 4]
```diff
@@ -60,7 +60,7 @@
     <Style Selector="Label.h3">
       <Setter Property="FontSize" Value="12" />
     </Style>
-  </Application.Styles>
+  </Application.Styles><!--
```
Comment (maxkatz6) [inline samples/ControlCatalog/App.xaml:63]: Is it supposed to be commented?

Comment (kekekeks) [reply samples/ControlCatalog/App.xaml:63]: I think it should be removed completely since it breaks control catalog on linux framebuffer. Which would make people to think that framebuffer platform is broken

File: src/Avalonia.Base/Utilities/DispatcherTimerHelper.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+namespace Avalonia.Utilities;
+
+public class DispatcherTimerHelper
+{
+    
+}
+
+public static class DispatcherTimerUtils
+{
+    public static void ForceFire(this DispatcherTimer timer)
```
Comment (maxkatz6) [inline src/Avalonia.Base/Utilities/DispatcherTimerHelper.cs:12]: Empty class.

Also, just for case, do we keep it public?
I think alternatively it could be internal here with public copy in Headless assembly. Either way, I am fine with it.

Comment (kekekeks) [reply src/Avalonia.Base/Utilities/DispatcherTimerHelper.cs:12]: It was supposed to go to Avalonia.Base.UnitTests, will fix

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032138-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]
